modules/subschema/src/main/scala/com.snowplowanalytics/iglu.schemaddl/jsonschema/subschema/package.scala

```diff
@@ -54,83 +63,109 @@ package object subschema {
       .getOrElse(s)
   }
 
+  @tailrec
+  def canonicalizeObject(s: Schema): Schema =
+    (s.properties, s.additionalProperties, s.patternProperties) match {
+      case (_, Some(AdditionalPropertiesAllowed(allowed)), _) =>
+        canonicalizeObject(s.copy(additionalProperties = Some(AdditionalPropertiesSchema(if (allowed) any else none))))
+      case (Some(Properties(props)), Some(AdditionalPropertiesSchema(_)), pProps) =>
+        val newPatternProps = props.map({case (k, v) => (s"^$k$$", v)}) ++ pProps.map(_.value).getOrElse(Map.empty)
+        // TODO: rewrite additionalProperties into patternProperties
+        s.copy(properties = None, patternProperties = Some(PatternProperties(newPatternProps)))
+      // TODO: handle overlapping patternProperties
+      case _ => s
+    }
+
+  def canonicalizeArray(s: Schema): Schema =
+    (s.items, s.additionalItems) match {
+      case (Some(ListItems(schema)), _) =>
+        s.copy(items = Some(TupleItems(List.empty)), additionalItems = Some(AdditionalItemsSchema(schema)))
+      case (_, Some(AdditionalItemsAllowed(false))) =>
+        s.copy(additionalItems = Some(AdditionalItemsSchema(none)))
+      case _ => s
+    }
```

Comment:
Should `AdditionalItemsAllowed(true)` also be transformed to empty Schema `any`?

Reply:
Good question. The canonicalization rule on the paper is specific to `false`, but I agree it makes sense in principle to also rewrite the `true` case. I'll have a deeper look as soon as I get to array subtype check

Reply:
I think the reason why `"additionalItems": true` true can't get rewritten to `"additionalItems": {}` is that it might not be right, depending on your usage of `items` (list vs tuple): if your array is an actual list of values and all items are supposed to validate against a single schema, rewriting `additionalItems` as a schema that matches anything would be wrong because it would contradict the fact that all the items in the array need to validate against the schema provided in `items`